feat: deduplicate with UMIs

mathiasbio opened the pull request:
### This PR:

Makes some significant changes to the TGA workflow by aiming to use the UMIs during the deduplication stage. This is a feature that only exists on the later versions of Sentieon. 

The PR is tested on Sentieon version:  sentieon-genomics-202308.03 which has this feature.

As this new feature was implemented it required some further changes to the workflows which ended up becoming a quite large refactoring. But in terms of effects on the quality of the analysis the only relevant changes are probably:
- using UMIs in the dedup for all TGA workflows
- adding adapter-trimming to the UMI workflow
- updating Sentieon version

Some more info regarding using the UMIs in the dedup step. It could rescue a significant number of reads wrongfully discarded as duplicates with a purely position-based approach, and could probably replace the UMI workflow for many use-cases as error-correction with 3,1,1 approach may not be necessary to avoid false positives in the range of VAFs commonly looked for in the samples previously run in the UMI workflow (0.5 - 2% VAF)

For more info see user story: https://github.com/Clinical-Genomics/BALSAMIC/issues/1361

### Further necessary features before merging this:

- [x] Replace Vardict with TNscope (OR verify that VarDict is performing better / same with these changes): https://github.com/Clinical-Genomics/BALSAMIC/pull/1429
- [ ] Above replaced by this PR: Merge VarDict and TNscope https://github.com/Clinical-Genomics/BALSAMIC/pull/1475
- [x] Rebuild TGA PONs with the new TGA UMI collapsed bams https://github.com/Clinical-Genomics/BALSAMIC/pull/1465


### Extra information:

**Extra note:**

- **Cap base quality script** was added because Manta crashes if it detects bases in the reads with a higher quality than 70, which some UMI reads get after the UMI collapse
- **samtools collate + samtools fixmate** was added to the TGA workflow because of bug in the new Sentieon dedup consensus tool 
- **awk script and samtools calmd added** to fix further issues in bamfile, documented here: https://github.com/Clinical-Genomics/BALSAMIC/issues/1472 and in the google sheet: https://docs.google.com/spreadsheets/d/1IZo1hcfYte2Vd7TjzQ-SjdX1wOqk9rHSNMQXvsDJbUM/edit?gid=573365273#gid=573365273

**Blocked by:**

- Update Sentieon: https://github.com/Clinical-Genomics/BALSAMIC/issues/1384
This update requires updating Sentieon as the current version in production: `sentieon-genomics-202010.02`
Does not contain the consensus option for LocusCollector which is the step prior to Dedup:

Current prod:
![image](https://github.com/Clinical-Genomics/BALSAMIC/assets/14308912/8af12b05-599b-4543-8a6e-d3c45702e665)

Latest Sentieon version:
![image](https://github.com/Clinical-Genomics/BALSAMIC/assets/14308912/c18ac360-0b2f-41d7-be27-a08f51382f2b)

**Issues to consider:**

- [x] Percent duplicate metrics:

After implementing this, we are no longer able to retrieve % Duplicates and Optical Duplicates info from the ".metrics" file from dedup. The values are "0". How can this be fixed? I have notified Sentieon about this issue and they say that they will add more useful statistics to the report in the next version. 

At the moment I have added standard non UMI Sentieon Dedup just to get the PERCENT DUPLICATES stats. But is it worth it? 

UPDATE: In the latest version of Sentieon 202308.03 they claim to have added support for % Duplicates from the --consensus mode from Dedup, so maybe this extra rule can be removed!

Comparing the percent duplicates between using Dedup with or without the UMI consensus options shows that they are a bit different, which makes sense as we're now using UMIs to identify duplicates as well as position. With roughly 4% less duplicates seen in this example case while using UMIs. Meaning that we're should be both getting a better calculation of duplicates which will be good for future evaluations of lab-methods, and at the same time saving more reads. 

|        | % duplicates dedup | % duplicates dedup --consensus |
|--------|--------------------|--------------------------------|
| normal |           0.540074 |                       0.481727 |
| tumor  |           0.569414 |                       0.510155 |

With this new version of sentieon we can safely remove the extra Sentieon dedup! 

- [x] Collapsed singleton or pairs?

Are overlapping mates in read-pairs maintained as Pairs or Singletons after UMI collapse (as in the other Sentieon UMI collapse tool)? I checked and they are singletons when the mates are overlapping. 

I discussed with Sentieon and colleagues and it should be fine: https://github.com/Clinical-Genomics/BALSAMIC/issues/1361 But it would be good to add more SVs from TGA to the validation set and make sure we can find them.

### Graph overview

**TGA T+N**

![image](https://github.com/Clinical-Genomics/BALSAMIC/assets/14308912/b0a3f8df-3321-4776-8a21-4d42b05b2dd3)


**TGA T+N (3,1,1 UMI workflow)**

![image](https://github.com/Clinical-Genomics/BALSAMIC/assets/14308912/50aa990f-f5e2-468b-93e3-012d9b7bd60d)


**TGA T+N (QC)**

![image](https://github.com/Clinical-Genomics/BALSAMIC/assets/14308912/a15b0101-b7ea-4298-b725-8281156f7deb)


**WGS T+N**

<img width="1108" alt="image" src="https://github.com/Clinical-Genomics/BALSAMIC/assets/14308912/6fb54f55-aaeb-4389-96e0-da192fa0474d">


**TGA CNVkit PON**

<img width="291" alt="image" src="https://github.com/Clinical-Genomics/BALSAMIC/assets/14308912/47cd7d17-306f-4126-8997-05dd60ff3cc2">



#### Added

- UMI extraction and deduplication to TGA workflow
- Adapter trimming of fastqs to UMI workflow
- Cap base quality in bam for Manta input

#### Changed

- Refactored multi workflow rule-files to separate files to decrease complexity
- Refactored output files to in general comply with format {sample_type}.{sample_name}
- Replaced Picard QC tools with matching Sentieon QC tools

#### Removed

- UMI specific rules for UMI-extraction and alignment (using new TGA-rules instead) 
- Fastq and UMI trimming command-line options

## Documentation

<!-- Link all added or updated documents. -->

- [ ] N/A
- [ ] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - [Document Name]

## Tests

<!-- Describe in detail how you tested your changes to help reviewers validate the code. -->
<!-- Include screenshots or visual representations of your changes. -->

### Feature Tests

<!-- Include tests relevant to the changes in this PR. -->

- [ ] All QC metrics are still detectable in deliverables yaml for TGA T+N and WGS T+N
- [ ] 

### Aggregate QC metric comparison to v15.0.0

Sheet with QC metric comparison:
https://docs.google.com/spreadsheets/d/1phztyRT1VS8ud6bf0JWwFo5PB2yWB0YgFXz_ZnXSF7U/edit?gid=0#gid=0

Notes on QC metrics:
> In general QC metrics are improved, but at the cost of somewhat longer run-times, increased by around 50%. This is due to the need of some extra rules for extracting UMIs and post-processing the bamfiles (the cap_base_qualities script and the Sentieon UMI extract rules, but there's no multi-threading option for the UMI extract). However hopefully this will be compensated by replacing VarDict with TNscope, we'll have to see, either way the TGA workflow is already quite fast so while it's a negative development it shouldn't be that bad. 

> On the positive side the QC metrics show a increase in coverage of between 2-20% after using the UMIs, which should increase the quality of the analysis and reduce the amount samples failing coverage thresholds in production.

> There's also a quite significant change in the number of variants called by VarDict in the TGA T/N case, of around 17%. But I will investigate this change in the associated PR where VarDict will be replaced by TNscope: https://github.com/Clinical-Genomics/BALSAMIC/pull/1429


- [x] QC metrics look good and we can proceed without further investigation.

### Pipeline Integrity Tests

<!-- Include tests to verify the integrity of the different Balsamic workflows. -->

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [x] N/A
    - [ ] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

## Clinical Genomics Stockholm

<details>
<summary></summary>

<!-- Do not reveal clinical data, and if applicable, place it within the internal Google Drive directory. -->

### Documentation

<!-- Link related issues or PRs for necessary changes. -->

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### User Changes

<!-- Please provide justification if you select N/A and the changes affect the output or results. -->

- [x] N/A
- [ ] This PR affects the output files or results.
    - [ ] User feedback is considered unnecessary because [Justification].
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

<!-- Link related issues or PRs for necessary changes. -->
<!-- Make sure that the linked PRs include relevant tests. -->

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

</details>

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [ ] Provided a comprehensive description of the PR.
    - [ ] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [ ] Verified and updated documentation if necessary.
- **Tests**
    - [ ] Described and tested the functionality addressed in the PR.
    - [ ] Ensured integration of the new code with existing workflows.
    - [ ] Confirmed that meaningful unit tests were added for the changes introduced.
    - [ ] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [ ] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [x] Code implements the intended features or fixes the reported issue.
    - [x] Code follows the project's coding standards and style guide.
- **Documentation**
    - [x] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Tests**
    - [x] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [ ] Confirmed that the developer has addressed all the comments during the code review.

<!-- Add any other relevant information or specific checks necessary for your PR. -->

ivadym commented on BALSAMIC/snakemake_rules/umi/sentieon_consensuscall.rule:
```diff
@@ -35,32 +35,32 @@ export SENTIEON_LICENSE={params.sentieon_lic};
 -o {output.fastq_consensus} \
 --input_format {params.ip_format} \
 --umi_tag {params.tag} \
 --read_name_prefix 'UMI-'; 
```
rm temp folder

ivadym commented on BALSAMIC/snakemake_rules/umi/sentieon_consensuscall.rule:
```diff
@@ -87,16 +87,16 @@ export SENTIEON_LICENSE={params.sentieon_lic};
         """
-# Filter consensus called reads based on 'XZ' filtering
 
+# Filter consensus called reads based on 'XZ' filtering
```
docstring

ivadym commented on BALSAMIC/snakemake_rules/umi/sentieon_consensuscall.rule:
```diff
@@ -35,32 +35,32 @@ export SENTIEON_LICENSE={params.sentieon_lic};
         """
-# Alignment of consensus reads
 
+# Alignment of consensus reads
```
docstring

ivadym commented on BALSAMIC/snakemake_rules/variant_calling/germline_tga.rule:
```diff
@@ -37,3 +37,38 @@ export SENTIEON_LICENSE={params.sentieon_lic};
         """
+
+
+rule manta_germline:
```
Could we have a single Manta germline run for both TGA and WGS, and then change the returned input BAM file based on the analysis type?

mathiasbio replied:
Yeah it's a good question. However it seems that for this function there won't be a default bam suffix that we could use for a given workflow that would work for all rules, so it would need to be determined by both rule and workflow name. In the end it sort of feels like we would be moving so much of the workflow logic into this get_bam_function 😅

mathiasbio replied:
For instance for manta_germline we would want the dedup.fixmate.qualcapped for the Manta TGA, but for the samtools_qc rule for tga we would want the align_sort

ivadym replied:
Let keep it simple then 🤷

It probably needs some proper refactoring of the general structure from the top instead

mathiasbio replied:
yeah! Agreed!

ivadym commented on BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal_tga.rule:
```diff
@@ -0,0 +1,166 @@
+rule manta_tumor_normal:
+    input:
+        fa = config["reference"]["reference_genome"],
+        bamN = config_model.get_final_bam_name(bam_dir=bam_dir, sample_name=normal_sample, specified_suffix="dedup.fixmate.qualcapped"),
+        bamT = config_model.get_final_bam_name(bam_dir=bam_dir, sample_name=tumor_sample, specified_suffix="dedup.fixmate.qualcapped"),
```
same as for  the`manta_germline` rule, the input seems  the only differences so it would nice if we can reuse the common code

ivadym replied:
Of course, if you consider this legacy code, we can ignore it and focus mainly on your PR

mathiasbio replied:
I partly like having multiple rules for different workflows even if they're practically identical, I think it's a bit more readable. It's clear from reading the file.rule name that the rules are active in the TGA workflow, you don't need to go into the rules constant and figure it out from there. I also kind of like to divide them up into workflows since the input data is so different 🤔 but maybe that's just my psychology :D

ivadym replied:
Yes that's a fair point. Maybe there is some middle ground solution to not have duplicated code, but not something we should discuss in this PR 😅

fevac replied:
For maintainability I think it's better to use the same rules unless there is a clear reason to not do this. Otherwise it's very easy for rules to evolve separately and fall out of sync and maintaining the code is trickier

mathiasbio replied:
I think one reason not to do it is that the input data is different, WGS vs TGA tend to require some different parameters. I'm almost more concerned that using the same rules for such different data-types makes it easier to miss that we might need to customise the rules. Such as we did before for the somatic version of manta where we didn't add the --exome option for the TGA. I don't think the risk is very high that we would consciously want to make some change to the manta analysis in for both workflows and that we miss doing it for both rules. But I don't have that strong opinions on this...

If I were to use 1 rule I don't know the best way to account for the input bam files needing to have different names for the two workflows. I could add some logic before like:

if sequencing_type == "wgs":
     bam_suffix="dedup.realign"
else:
     bam_suffix="dedup.fixmate.qualcapped"

Would you prefer something like that?

ivadym replied:
From what I see, it's not possible to add these conditions to the `get_final_bam_name` method, right? 

If it works to have this logic in the `.rule`, then sure, it's probably better to have one rule for better maintainability, as Eva suggested

mathiasbio replied:
We're in the same situation for a few other rules too, for example the somatic manta rule. I divided up in general some `.rule` files that were shared between the different workflows to avoid having to read these different if-statements to understand which rules are used by which workflows, which I think was really messy to read. But the downside of dividing them up into workflows is that we get these duplicated rules. Another instance is for the somatic manta variant-calling which is now also practically identical. But I don't see any way around it now without re-introducing some of these if-statements... I think in general it's just easier to find what you're looking for when you just need to open the .rule file with the "tga" or "wgs" in the name. If I'm going to change this for this rule I should probably be consistent and change it back for all the others too, but there's going to be a lot of if-statements in the rules...and I feel like it will be messier...I think like Vadym said earlier we can think of a better structure in a later refactoring 🤷

ivadym commented on BALSAMIC/assets/scripts/collect_qc_metrics.py:
Should we re-evaluate the QC thresholds, or do you think it's not relevant?

mathiasbio replied:
In general it doesn't look like anything really relevant has changed for our deliverables except the expected increase in coverage. Which should result in fewer QC fails, but I don't think it should warrant changing any threshold. There's also a mild decrease in GC_DROPOUT but not really noticeable. So probably we're fine here : )

ivadym replied:
Nice, thanks for looking into it 🙏  I will resolve and we monitor the analyses in production. If there are less or more QC fails they will let us know 😄

ivadym reviewed (commented):
Fantastic work! Amazing effort in adapting all the workflows! 💯 🦾 🎖️

mathiasbio commented on BALSAMIC/assets/scripts/cap_base_quality_in_bam.py:
```diff
@@ -0,0 +1,35 @@
+    with pysam.AlignmentFile(input_bam, "rb") as in_bam:
+        # Open output BAM file for writing
+        with pysam.AlignmentFile(output_bam, "wb", header=in_bam.header) as out_bam:
+            # Iterate over reads in the input BAM file
+            for read in in_bam:
+                # Cap base qualities to the maximum value
+                capped_qualities = [min(q, max_quality) for q in read.query_qualities]
+                # Update the base qualities in the read
+                read.query_qualities = capped_qualities
+                # Write the modified read to the output BAM file
+                out_bam.write(read)
```
unitedbeagle took about 20 minutes, selectbengal which is much larger took 2 hours. So it can take some time...it is however a parallel step which only leads to manta variant-calling and from the case I looked at it seems that the SNV workflow still finished after the SV workflow so it doesn't seem to be delaying that TAT. But if it could be faster that would still be nice...

ivadym replied:
Yes... but I couldn't find anything. They use `fetch()` without arguments in the documentation and then iterate over it, but I'm not sure it will make any difference. Feel free to resolve this :)

mathiasbio replied:
I will test it and see if it's faster!

mathiasbio replied:
I ran the same bam file with the old method, and then with fetch, and it went from  2:46.16 to 2:37.14, so maybe it's a bit faster. I'll try and see if I can find any other way to make it faster

ivadym replied:
I think the main problem is the for loop, but I'm not sure you can vectorise the output from fetch() with numpy for example, and then call [numpy.minimum](https://numpy.org/doc/stable/reference/generated/numpy.minimum.html#numpy-minimum).  If that's possible it should be more efficient, but it seems what we get is kind of a dictionary. Not sure I would invest much time on this but amazing if you make it work 🤩

mathiasbio replied:
I tried parallelising it but it was a mess 😅 but it seems that the numpy solution is working! It went to: 1:31.34 so that's a pretty huge improvement. Thanks for the tip!

mathiasbio commented on BALSAMIC/commands/config/case.py:
```diff
@@ -89,19 +84,15 @@
 @OPTION_SWEGEN_SNV
 @OPTION_SWEGEN_SV
 @OPTION_TUMOR_SAMPLE_NAME
-@OPTION_UMI
-@OPTION_UMI_TRIM_LENGTH
```
Yeah! If you want I can try to add the functionality back to optionally trim the UMIs. It could be nice for external users, or if we want to stop using UMIs, but I doubt either case is very important or likely to occur 🤔 If you don't think the functionality is important I'll open issues in CG!

ivadym replied:
I don't think the effort is worth it at this point, we can just remove it here and in CG

mathiasbio replied:
I made an issue in CG and linked to the release 16 PR: https://github.com/Clinical-Genomics/cg/issues/3416

ivadym replied:
Beautiful 🌹

mathiasbio commented on BALSAMIC/workflows/QC.smk:
```diff
@@ -57,6 +57,9 @@ vcf_dir: str = Path(result_dir, "vcf").as_posix() + "/"
 qc_dir: str = Path(result_dir, "qc").as_posix() + "/"
 delivery_dir: str = Path(result_dir, "delivery").as_posix() + "/"
 
+# Pre run parameters
+seconds_before_start: int = SLEEP_BEFORE_START
+
```
I think we are using it, and it also makes sense to use it I think due to the key error, in all workflows right? 🤔 Or should it not be an issue in smaller workflows like the QC because it should finish submitting all the jobs in time? Either way I guess it doesn't hurt to sleep a few minutes

ivadym replied:
I think we're using directly `SLEEP_BEFORE_START`, not `seconds_before_start`. We should not redefine constants, that's why I though it was some kind of code conflicts between your other branches

mathiasbio replied:
Ahh good point. I'll fix it 👍

mathiasbio commented on BALSAMIC/workflows/PON.smk:
```diff
@@ -49,6 +49,9 @@ bam_dir: str = Path(result_dir, "bam", "").as_posix() + "/"
 cnv_dir: str = Path(result_dir, "cnv", "").as_posix() + "/"
 qc_dir: str = Path(result_dir, "qc", "").as_posix() + "/"
 
+# Pre run parameters
+seconds_before_start: int = SLEEP_BEFORE_START
+
```
Same here, we're using the same rules as we're using in the balsamic.smk and they rely on the output from this sleep rule. And probably especially in the PON workflow it's required to sleep due to how many jobs we're submitting from here when we're running like 50 samples or something :D

mathiasbio commented on BALSAMIC/commands/config/case.py:
```diff
@@ -198,19 +186,14 @@ def case_config(
         directory.mkdir(exist_ok=True)
 
     config_collection_dict = ConfigModel(
+        QC={},
```
Thanks! I agree that looks a lot nicer

mathiasbio commented on BALSAMIC/snakemake_rules/quality_control/fastp_tga.rule:
```diff
@@ -35,50 +34,78 @@ fastp \
+
+mkdir -p {params.tmpdir};
+export TMPDIR={params.tmpdir};
+export SENTIEON_TMPDIR={params.tmpdir};
+export SENTIEON_LICENSE={params.sentieon_lic};
```
Yeah... 🤔 maybe we can add it as a future refactoring though? I took a brief look and I couldn't quite figure it out, and for the TMPDIR variables I still expect that they would need to be set in the rules themselves since it's a dynamic variable that changes values for each rule

ivadym replied:
Oh, sorry, I only meant the Sentieon license directory, which is common and immutable. But sure, this PR is already quite large

mathiasbio commented on BALSAMIC/snakemake_rules/quality_control/fastp_tga.rule:
```diff
@@ -35,50 +34,78 @@ fastp \
---disable_trim_poly_g \
-{params.fastp_trim_umi};
+--overrepresentation_analysis \
+{params.adapter_trim}
```
I think so! I'll add the rm -rf command wherever I see the tmpfolder has been created

mathiasbio commented on BALSAMIC/snakemake_rules/align/tga_bam_postprocess.rule:
```diff
@@ -0,0 +1,103 @@
+        "Capping base qualities to 70 in bamfile using pysam in sample: {params.sample_id}"
+    shell:
+        """
+python {params.cap_base_qualities} --max_quality 70 {input.bam} {output.bam} ; 
```
Added a BAMPostProcessingParams model

mathiasbio commented on BALSAMIC/snakemake_rules/quality_control/sentieon_qc_metrics.rule:
```diff
@@ -9,49 +9,173 @@ def repeat(param, values):
+            Path(benchmark_dir,'sentieon_wgs_metrics_' + "{sample_type}_{sample}.tsv").as_posix()
+        params:
+            tmpdir = tempfile.mkdtemp(prefix=tmp_dir),
+            min_base_qual = '10',
```
There's quite a lot of these constants in the QC rules 😓

ivadym replied:
😓 

Fix the ones that are manageable. It's a general problem we have in our codebase. Similar to the SV related rules, they all have hardcoded values 🙃

mathiasbio replied:
I'm converting some of them into models at least 👍

mathiasbio commented on BALSAMIC/snakemake_rules/quality_control/sentieon_qc_metrics.rule:
```diff
@@ -9,49 +9,173 @@ def repeat(param, values):
+{params.sentieon_exec} plot InsertSizeMetricAlgo {input.insert_size_metrics} -o {output.insert_size_pdf} ; 
+{params.sentieon_exec} plot GCBias {input.gc_bias_metrics} -o {output.gc_bias_pdf} ; 
+{params.sentieon_exec} plot MeanQualityByCycle {input.quality_by_cycle_metrics} -o {output.quality_by_cycle_pdf} ; 
+{params.sentieon_exec} plot QualDistribution {input.quality_distribution_metrics} -o {output.quality_distribution_pdf} ; 
```
Yeah me neither, it's just a small plotting job so probably no temp folder required even. Maybe I'll just remove the create tmp from params

ivadym replied:
yes sounds good

fevac commented:
Hej, I don't understand this from the description:

>  Collapsed singleton or pairs?
Are the final collapsed reads maintained as Pairs or Singletons (as in the other Sentieon UMI collapse tool)?
They are singletons, and Sentieon says this is intentional and beneficial.

Do you mean clusters or read-pairs? I'm a bit confused

mathiasbio commented:
@fevac Yeah it's a bit unclear :D some UMI tools when collapsing overlapping mates maintain the reads read-pairs but in the Sentieon tools if the mates are overlapping they are collapsed into singletons. I knew that the UMI tools in the UMI workflow works like this, but the --consensus option in Dedup is a new feature that we're activating in the TGA workflow and I wasn't sure how it handled overlapping mates, and I wasn't sure what was the best solution. So I discussed a bit with Khurram and Jesper, and emailed Sentieon about it (there's a little more info here: https://github.com/Clinical-Genomics/BALSAMIC/issues/1361 ) and at least according to them, creating singletons out of such read-pairs is fine / better than keeping them as pairs. I'm a bit concerned still about the effects on SV-calling, in theory it should be fine, but I still think we should add some true SVs for the TGA validation

mathiasbio commented on BALSAMIC/models/params.py:
```diff
@@ -37,6 +45,41 @@ class ParamsManta(BaseModel):
+class ParamsMosdepth(BaseModel):
+    """This class defines the params settings used as constants in Mosdepth rule.
+
+    Attributes:
+        mapq: str(required); mapping quality threshold, reads with a quality less than this value are ignored [default: 0]
+        samflag: str(required); exclude reads with any of the bits in FLAG set [default: 1796]
+        quantize: str(required); merges adjacent bases as long as they fall in the same coverage bins e.g. (10-20)
```
these default values are the default from the tools, I just did "--help" because I wasn't sure what the parameters meant. I figured it could be a bit nice to know that we're changing the default from the tool but it's the wrong place for this info 🤔 either way I'll remove the default text

mathiasbio commented on BALSAMIC/snakemake_rules/umi/sentieon_consensuscall.rule:
```diff
@@ -69,13 +69,12 @@ rule sentieon_bwa_umiconsensus:
 
 {params.sentieon_exec} bwa mem \
--R {params.sheader} \
+-R '@RG\\tID:{params.sample_id}\\tSM:{params.sample_type}\\tLB:TargetPanel\\tPL:ILLUMINA' \
```
I honestly don't remember 🤔  I took a look at the bamfiles produced in production from the UMI pipeline and it looks like the params.sheader is working. Could be that I removed it to be consistent as we're not using it for other alignment rules. But if we wanted to add it we would need to add 2 versions of the header, one for the one where the is the fastq_pattern (per lane) and one where the alignment is per sample. In the end I think it's kind of nice to be able to see it directly in the rule but I have no strong opinions. I can add the two headers if you prefer it : )

ivadym replied:
For me it would be clearer if we use params, which will make it easier to check from the name which header is used for which rule or workflow

ivadym replied:
And if we change it in the future for some reason, or update it, then it's just in one place

mathiasbio replied:
Alright! I'll add 🎈

mathiasbio replied:
I realize I cannot use the params because the sample_type we're using in the bamfile should be capitalised TUMOR or NORMAL whereas the wildcards for sample_type that we have access to in the model is not. So I don't know how simple it would be. Maybe we can refactor in the future

ivadym replied:
Oh that's annoying, yes, and there doesn't seem to be any snakemake directives to capitalize params... Let's keep it hardcoded then

ivadym commented on BALSAMIC/assets/scripts/cap_base_quality_in_bam.py:
```diff
@@ -0,0 +1,38 @@
+import click
+import pysam
+import numpy as np
+from BALSAMIC.constants.workflow_params import WORKFLOW_PARAMS
```
seeing this made me remember that we don't have balsamic installed inside that container, so actually it won't find the WORKFLOW_PARAMS constant

sorry for confusing you 🙏  leaving the hardcoded default value should be ok

mathiasbio replied:
Impressive that you realized :p quick fix!

ivadym commented on BALSAMIC/snakemake_rules/align/tga_sentieon_alignment.rule:
```diff
@@ -0,0 +1,44 @@
+{params.sentieon_exec} bwa mem \
+-R {params.header} \
+-K {params.ip_bases} \
+-t {threads} -C \
```
What's this `-C` flag? I couldn't find in it the documentation

ivadym commented on BALSAMIC/snakemake_rules/align/wgs_bam_postprocess.rule:
```diff
@@ -107,18 +65,77 @@ rule sentieon_realign:
+-o {output.qual_recal_plot} {output.qual_recal};
 
+rm -rf {params.tmpdir};
+        """
```
EOF
```suggestion
        """

```

ivadym reviewed (approved):
Codewise looks great, as well as the results! 💯 Amazing work and fantastic documentation of the process! 🤩

mathiasbio commented:
Thanks for the reviews! As discussed I'll not merge this however until I have fully tested the replace VarDict PR, and before this PR is entirely complete I'll still need to add some documentation 🙏

mathiasbio commented:
Merged this PR into this one: https://github.com/Clinical-Genomics/BALSAMIC/pull/1465

#### Added

- Added extension of target bed regions to a minimum size of 100 for CNV analysis
- PON for: Exome comprehensive 10.2 
- PON for: GMSsolid 15.2 
- PON for: GMCKsolid 4.2

#### Changed

- updated PON for GMCKSolid v4.1 
- updated PON for GMSMyeloid v5.3 
- updated PON for GMSlymphoid v7.3

Merged this PR into this one: https://github.com/Clinical-Genomics/BALSAMIC/pull/1448

#### Added

- Script to post-process CNVkit output to GENS-format
- DNAscope gnomad calling to TGA for GENS

#### Changed

- Parsing of GENS arguments changed to account for TGA

mathiasbio commented:
Merged this PR: https://github.com/Clinical-Genomics/BALSAMIC/pull/1475 into this one

#### Changed

- Refactored rules for bcftools filters
- Renamed final UMI bamfile to ensure hsmetrics are collected in multiqc json
- Changed ranked VCF from research to clincial
- Lowered min AF for TGA from 0.007 to 0.005
- Lowered maximal SOR for TNscope in TGA tumor only cases from 3 to 2.7
- Changed filter settings for research TNscope vcf, now either PASS or triallelic_site (fixing this issue: https://github.com/Clinical-Genomics/BALSAMIC/issues/1293)

#### Added

- TNscope for TGA workflows, merged with VarDict results
- New filter for VarDict for tumor in normal contamination
- Export TMP environment variables to rules that lack them
- Added genmod ranked VCFs to be delivered
- Added family-id to genmod in order to get ranked variants to Scout (solved this: https://github.com/Clinical-Genomics/BALSAMIC/issues/1045)
- Added DP and AF to INFO-field of TNscope vcfs for ranking model
- Raw TNscope calls and unfiltered research-annotated SNVs to delivery

#### Removed

- ML-model for TNscope is removed due to license issue with new version of Sentieon
- All code associated with TNhaplotyper
- Removed research.filtered.pass VCFs from delivery and storage list

mathiasbio commented on BALSAMIC/models/config.py:
```diff
@@ -434,11 +439,14 @@ def get_final_bam_name(
         elif self.analysis.sequencing_type == SequencingType.TARGETED:
             # Only dedup is necessary for TGA
-            final_bam_suffix = "dedup_sorted"
+            final_bam_suffix = "dedup.fixmate"
         else:
             # For WGS the bamfiles are realigned
             final_bam_suffix = "dedup.realign"
 
```
At least the bamfiles that we deliver to the customers are the cram-files and they are all named the same across all the workflows regardless of which bamfile was compressed. So these final_bam_suffixes are in the end all renamed to the same cram-file : )

mathiasbio commented on BALSAMIC/constants/rules.py:
```diff
@@ -66,22 +64,23 @@
             "snakemake_rules/variant_calling/somatic_cnv_tumor_only_tga.rule",
             "snakemake_rules/variant_calling/somatic_tumor_only.rule",
-            "snakemake_rules/variant_calling/somatic_sv_tumor_only.rule",
+            "snakemake_rules/variant_calling/somatic_sv_tumor_only_tga.rule",
```
I'm sure that a better structure can be made! Let's work on it for the next release 💪 🙏

mathiasbio commented:
Thanks for the review! ❤️
